lib/segment/src/index/field_index/geo_index.rs

```diff
@@ -1 +1,533 @@
+        // traverse tiles matrix by columns - top to bottom
+        for _ in 0..width {
+            let mut current = top.clone();
+            for _ in 0..height {
+                seen.push(current.clone());
+                current = neighbor(&current, Direction::S).unwrap();
+            }
+            // move top column to the east
+            top = neighbor(&top, Direction::E).unwrap();
+        }
+        seen
+    }
+
+    fn area_width_region_count(&self) -> usize {
+        let mut width_region = 0;
+        // count starting tile
+        if self.north_west != self.north_east {
+            width_region += 1;
+        }
+        let mut current = self.north_west.clone();
+        while current != self.north_east {
+            let east_neighbor = neighbor(&current, Direction::E).unwrap();
+            current = east_neighbor;
+            width_region += 1;
+        }
+        width_region
+    }
+
+    fn area_height_region_count(&self) -> usize {
+        let mut height_region = 0;
+        // count starting tile
+        if self.north_west != self.south_west {
+            height_region += 1;
+        }
+        let mut current = self.north_west.clone();
+        while current != self.south_west {
+            let south_neighbor = neighbor(&current, Direction::S).unwrap();
+            current = south_neighbor;
+            height_region += 1;
+        }
+        height_region
+    }
+
+    /// number of geohashes within the rectangle
+    fn area_region_count(&self) -> usize {
+        self.area_height_region_count() * self.area_width_region_count()
+    }
+
+    fn compute_from_bounding_box(bounding_box: &GeoBoundingBox, geohash_precision: usize) -> Self {
+        let GeoPoint {
+            lat: max_lat,
+            lon: min_lon,
+        } = bounding_box.top_left;
+        let GeoPoint {
+            lat: min_lat,
+            lon: max_lon,
+        } = bounding_box.bottom_right;
+
+        // Unwrap is acceptable, as data should be validated before
+        let north_west = encode((min_lon, max_lat).into(), geohash_precision).unwrap();
+        let south_west = encode((min_lon, min_lat).into(), geohash_precision).unwrap();
+        let south_east = encode((max_lon, min_lat).into(), geohash_precision).unwrap();
+        let north_east = encode((max_lon, max_lat).into(), geohash_precision).unwrap();
+
+        Self {
+            north_west,
+            south_west,
+            south_east,
+            north_east,
+        }
+    }
+}
+
+/// Return as-high-as-possible with maximum of `max_regions`
+/// number of geo-hash guaranteed to contain the whole circle.
+fn circle_hashes(circle: GeoRadius, max_regions: usize) -> Vec<GeoHash> {
+    let rectangle = minimum_bounding_rectangle_for_circle(&circle);
+    let inner_hashes = rectangle_hashes(rectangle, max_regions);
+    filter_hashes_within_circle(inner_hashes, circle)
+}
+
+/// filter geo hashes for which the center is within a circle
```

Comment:
If we only consider center of the tile - some parts of the circle could be not covered by resulting hashes. In this case it might be better to just drop this filtering and rely on exact post-filtering.

Or the alternative - to exclude only those tiles which are more distant than radius + half of the diagonal.

Reply:
![IMG_1324](https://user-images.githubusercontent.com/1935623/154752721-cae17e31-1d42-43cb-b940-4b73300a6027.PNG)

Reply:
higher precision might create huge mess with trigonometry and easily take another day of work